### PR TITLE
Viscous stress and ellipse ib

### PR DESCRIPTION
**Pull request by danieljvickers**
## Description

I added code to handle two-way coupling for moving IBMs with viscous forces. I also added an ellipsoidal IB patch for sediment-based cases that we may want to run. I also have added a new example case of the 2D shock cylinder, which was analyzed and compared to a published example from "Moving overlapping grids with adaptive mesh refinement for high-speed reactive and non-reactive flow" by William D. Henshaw and Donald W. Schwendeman. The position, force, and velocity of that case is shown below

full

Of note, there are a few things missing from this PR is higher-order derivatives, which is going to come in a more in-depth validation PR next. I also have checked that the results make sense in 3D with a moving sphere case, but I have not done rigorous analysis of a published 3D case yet.

The code already works with MPI and GPU capability, and was adequately tested on phoenix and frontier for both features. For performance, I gathered some grind time numbers for 3 different IB cases on Frontier. These cases include a stationary IB with a moving fluid, a stationary fluid with a moving IB, and an inviscid stationary fluid with moving IB. Those times are shown here

image

### Type of change

- [ ] New feature (non-breaking change which adds functionality)

### Scope

- [ ] This PR comprises a set of related changes with a common goal

If you cannot check the above box, please split your PR into multiple PRs that each have a common goal.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Please also list any relevant details for your test configuration

- [ ] I ran a comparison to a published shock-cylinder case, as well as verification for things like boyant forces and verification that it works with gravity.
- [ ] I tested locally with GNU as well as on Phoenix with NVHPC on GPUs and obtained identical results on both.

**Test Configuration**:

* What computers and compilers did you use to test this:

## Checklist

- [ ] I have added comments for the new code
- [ ] I added Doxygen docstrings to the new code
- [ ] I have made corresponding changes to the documentation (`docs/`)
- [ ] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [ ] I have added example cases in `examples/` that demonstrate my new feature performing as expected.
They run to completion and demonstrate "interesting physics"
- [ ] I ran `./mfc.sh format` before committing my code
- [ ] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [ ] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [ ] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [ ] Checked that the code compiles using NVHPC compilers
- [ ] Checked that the code compiles using CRAY compilers
- [ ] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [ ] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [ ] Enclosed the new feature via `nvtx` ranges so that they can be identified in profiles
- [ ] Ran a Nsight Systems profile using `./mfc.sh run XXXX --gpu -t simulation --nsys`, and have attached the output file (`.nsys-rep`) and plain text results to this PR
- [ ] Ran a Rocprof Systems profile using `./mfc.sh run XXXX --gpu -t simulation --rsys --hip-trace`, and have attached the output file and plain text results to this PR.
- [ ] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added viscous stress tensor computation for moving immersed boundaries with two-way coupling

- Implemented ellipsoidal immersed boundary patch geometry for 2D simulations

- Extended force and torque calculations to include viscous stress contributions

- Added body force support (gravity) to immersed boundary dynamics

- Included shock-cylinder and ellipse example cases with validation tests

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Viscous Stress Tensor"] --> B["Force/Torque Calculation"]
  C["Ellipse IB Patch"] --> D["Levelset Generation"]
  B --> E["Two-Way Coupling"]
  D --> E
  F["Body Forces"] --> E
  E --> G["Moving IBM Dynamics"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>m_ibm.fpp</strong><dd><code>Integrated viscous forces and body forces into IBM</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-5d003cc844bee6db37016a0c170bde96166b6267daf701d1b78b4fc51bc179da">+118/-46</a></td>

</tr>

<tr>
  <td><strong>m_viscous.fpp</strong><dd><code>Added viscous stress tensor computation routine</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-238582481ed5a2e2ce1f912a532cdb3f520d024304c931d3e5e23b724f64eb31">+69/-6</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>m_compute_levelset.fpp</strong><dd><code>Implemented ellipse levelset generation method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-01515391f4d51c0514aa4d25c7e79021b01721edb7577cac7b87a08eb8144739">+58/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>m_ib_patches.fpp</strong><dd><code>Added ellipse immersed boundary patch implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-a3aa8375918913e3364ba1f5fff6fd5e73e835ee3524ef2a32a27c4177b8aa25">+44/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>m_time_steppers.fpp</strong><dd><code>Updated force computation call with viscosity parameter</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-be32e0928c6a2e0682fbb3de74a580b048cd57abbb5899b65bd1278661d560a8">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>m_check_ib_patches.fpp</strong><dd><code>Added validation check for ellipse patch geometry</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-db84405c241cb20f276f26b7d50e7cf410dccdcf46ec7114d7850046347f9a4c">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>m_rhs.fpp</strong><dd><code>Renamed viscous stress function for clarity</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-fe8ed9906ee74f3d3f0dbc702d75863fc90286a345cc7067acc6e6cfe80b6d52">+12/-12</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>case.py</strong><dd><code>New shock-cylinder test case with moving cylinder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-30953f2918ec55adbdfd25105d6c8bcf905132655cc70cc7b2e0002f4e4520b4">+138/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>case.py</strong><dd><code>New ellipse immersed boundary example case</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-9c9137fe65a918e7ecd4f6e0cbe91d517e62593702dbbff11655b6287a98566c">+97/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>golden-metadata.txt</strong><dd><code>Golden test metadata for shock-cylinder case</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-f0d9a48e1a126a1b79083bcfbea990c0ffde3ef34384fdf816ddf75102e2764a">+193/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>golden-metadata.txt</strong><dd><code>Golden test metadata for ellipse case</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-2a7d13cb58542aee07aab83f1acd3aacdb3ae855d1fd5f93f96da51778f10236">+193/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>golden.txt</strong></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-5fbed276e4d2b03a3e84a6c05440dc11ec83f4b8e6570155898cf2c024985fe9">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>golden.txt</strong></td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/1102/files#diff-dd846d2acb6b47140d9b8658a624e41e1969bc9768a25d976bed4734b216c851">+11/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

___

## **CodeAnt-AI Description**
Add viscous two-way coupling for moving immersed boundaries and ellipse IB support

### What Changed
- Two-way coupling now includes viscous stresses: moving immersed boundaries receive forces and torques from both pressure and viscous stress contributions, and those summed forces update IB linear and angular motion.
- Elliptical immersed boundary geometry added: domain markers, levelset normals, and levelset values are computed for ellipse patches so ellipsoidal bodies can be used in simulations.
- Ghost-cell treatment and IB initialization updated: interior ghost cells for moving IBs set momentum consistent with the IB velocity so moving bodies start with correct local momentum; moment-of-inertia computation corrected for ellipse geometry.
- Body forces (gravity/acceleration) are applied to IBs after global reduction to avoid double counting.
- Two new example case files added: a 2D shock-cylinder case and a 2D IB ellipse case demonstrating viscous and moving-IB behavior.

### Impact
`✅ Two-way viscous coupling for moving bodies`
`✅ Simulations can include ellipsoidal immersed boundaries`
`✅ Examples available to reproduce viscous moving-IB scenarios`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **New Features**
  * Elliptical immersed-boundary support
  * Viscous coupling for immersed-boundary forces and improved moving-body inertia handling
  * New example case generators that print JSON configs for 2D ellipse and 2D shock-cylinder

* **Bug Fixes**
  * Corrected viscous-stress handling for cylindrical-boundary cases and fixed related argument usage

* **Tests**
  * Added extensive golden-data files and run metadata for new scenarios

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/simulation/m_ibm.fpp [hunk 1]
```diff
@@ -997,62 +1012,104 @@ contains
...
-                                $:GPU_ATOMIC(atomic='update')
-                                forces(ib_idx, l) = forces(ib_idx, l) - (pressure_divergence(l)*cell_volume)
-                                $:GPU_ATOMIC(atomic='update')
-                                torques(ib_idx, l) = torques(ib_idx, l) - local_torque_contribution(l)
+                                call s_cross_product(radial_vector, viscous_stress_div_1(l, 1:3), viscous_stress_div_1(l, 1:3))
+                                call s_cross_product(radial_vector, viscous_stress_div_2(l, 1:3), viscous_stress_div_2(l, 1:3))
                             end do
+
+                            viscous_stress_div = (viscous_stress_div_2 - viscous_stress_div_1)/(2._wp*dy)
+                            local_torque_contribution(1:3) = local_torque_contribution(1:3) + viscous_stress_div(2, 1:3)
```
[the remark this answers is not in the document]
Comment (danieljvickers) [reply src/simulation/m_ibm.fpp:1091]: Objectively wrong. Ignoring.

**Comment by danieljvickers** [comment]
Spencer, the phoenix benchmark is failing because pip is not installing python dependencies correctly. Seems like an issue outside of the scope of files changes and more a  problem with master. What do you think.

**Comment by sbryngelson** [comment]
just needed to be restarted. @danieljvickers can you close the ai reviewer comments that are irrelevant/false positives?

**Comment by danieljvickers** [comment]
I resolved them all

File: src/simulation/m_time_steppers.fpp [hunk 2]
```diff
@@ -627,7 +627,7 @@ contains
 
                             if (patch_ib(i)%moving_ibm == 2) then ! if we are using two-way coupling, apply force and torque
                                 ! compute the force and torque on the IB from the fluid
-                                call s_compute_ib_forces(q_prim_vf(E_idx))
+                                call s_compute_ib_forces(q_prim_vf, 1._wp/fluid_pp(1)%Re(1))
```
Comment (sbryngelson) [inline src/simulation/m_time_steppers.fpp:630]: This blows up at Re = 0 (or no Re). Also it seems like all of this should be abstracted to a subroutine

File: src/simulation/m_viscous.fpp [hunk 3]
```diff
@@ -1511,6 +1512,68 @@ contains
 
     end subroutine s_compute_fd_gradient
 
+    ! computes the viscous stress tensor at a particule i, j, k element
+    subroutine s_compute_viscous_stress_tensor(viscous_stress_tensor, q_prim_vf, dynamic_viscosity, i, j, k)
```
Comment (sbryngelson) [inline src/simulation/m_viscous.fpp:1516]: I'm a bit confused. What happened here? Maybe the diff is just weird.

File: src/simulation/m_time_steppers.fpp (hunk 2 above)
Comment (danieljvickers) [reply src/simulation/m_time_steppers.fpp:630]: This runs in inviscid cases (I have confirmed this) because the Reynolds number is set to the `dflt_real` value, which is non-zero, in those cases. You are correct though that if someone was running with viscosity and set the dynamic viscosity to infinity (by setting Re to 0) that this would break. There is also no check going on in preprocessing that checks for this, s I didn't think to check that.

I think a more-sustainable solution would be to add a prohibit statement for viscous cases using Re=0, because I don't think that problem is well posed. Please correct me on this if I am mistaken.

File: src/simulation/m_viscous.fpp (hunk 3 above)
Comment (danieljvickers) [reply src/simulation/m_viscous.fpp:1516]: There was a subroutine that  was called `s_compute_viscous_stress_tensor`. It just corrected the viscous stress tensor near the boundaries in cylindrical coordinates. There was no subroutine that directly computed this value at cell centers. I had to add it in order to get these values for particular grid cells to compute the viscous force. So I renamed the old routine to be more explicit about what it actually did, and the added a subroutine to compute the viscous stress tensor.
